processing/src/main/java/io/druid/query/groupby/having/DimFilterHavingSpec.java

```diff
@@ -54,19 +57,24 @@ public DimFilter getDimFilter()
   }
 
   @Override
-  public boolean eval(final Row row)
+  public void setRowType(Map<String, ValueType> rowType)
   {
-    // Not thread safe, but it doesn't have to be.
-    valueMatcherFactory.setRow(row);
-    try {
-      return valueMatcher.matches();
-    }
-    finally {
-      valueMatcherFactory.setRow(null);
+    synchronized (valueMatcherMonitor) {
```

Comment:
I'm also adding a detector for concurrent usage just in case, so this will fail fast if anyone breaks the contract in the future.